[5.2] Cache based sessions don't need to take part in the GC lottery

HolgerW1 opened the pull request:
Cache based session handlers, `CacheBasedSessionHandler` in particular, don't support the garbage collection used by the `StartSession` middleware.
So there's actually no need for them to take part in the GC lottery or to call the the `gc` method which doesn't have any effect.

Commit:
Sessions using cache based handlers don't need to take part in the garbage collection lottery and don't need to trigger the garbage collection

GrahamCampbell commented on src/Illuminate/Session/Middleware/StartSession.php:
```diff
@@ -8,8 +8,9 @@
 use Illuminate\Session\SessionInterface;
-use Symfony\Component\HttpFoundation\Cookie;
 use Illuminate\Session\CookieSessionHandler;
+use Illuminate\Session\CacheBasedSessionHandler;
```
Please sort be length.

HolgerW1 replied:
Like this ? (see latest commit)
Couldn't find any coding style documentation regarding this. Wouldn't it be better to sort via namespace? I mean I don't really care but it would be great if you would document those CS guidelines somewhere. Or if StyleCI would automatically check for this.

GrahamCampbell replied:
> Or if StyleCI would automatically check for this.

It doesn't fix that yet.

Commit:
Include order changed

GrahamCampbell commented on src/Illuminate/Session/Middleware/StartSession.php:
```diff
@@ -243,4 +249,18 @@ protected function usingCookieSessions()
+            return false;
+        }
+
+        return $this->manager->driver()->getHandler() instanceof CacheBasedSessionHandler;
```
this code is broken

GrahamCampbell replied:
you need to pass through the correct `$session` instance

HolgerW1 replied:
> you need to pass through the correct $session instance

You're right. Should be fixed now.

Commit:
Fixed check for session based handler

GrahamCampbell commented on src/Illuminate/Session/Middleware/StartSession.php:
```diff
@@ -243,4 +249,15 @@ protected function usingCookieSessions()
 
         return $this->manager->driver()->getHandler() instanceof CookieSessionHandler;
```
this code is wrong i guess

GrahamCampbell replied:
unrelated to the PR though

GrahamCampbell commented on src/Illuminate/Session/Middleware/StartSession.php:
```diff
@@ -143,13 +144,18 @@ protected function storeCurrentUrl(Request $request, $session)
+        // Cache based sessions use their own mechanism to invalidate entries
+        // once their lifetime expires so there's no need to perform the garbage
+        // collection for them.
+        if (! $this->sessionUsesCacheBasedHandler($session)) {
```
Better to early return and not mess up the nesting

GrahamCampbell replied:
```php
if ($this->sessionUsesCacheBasedHandler($session)) {
    return;
}
````

HolgerW1 replied:
I agree, what about changing the name of the method to `sessionHandlerSupportsGc` before I push another commit? I.e. a generic function which checks if a session handler type supports garbage collection and returns false for cache based handlers.

GrahamCampbell replied:
Supports isn't really correct. They both support it, just not all of them need it.

HolgerW1 replied:
Okay, `sessionHandlerNeedsGc` then?

GrahamCampbell replied:
What about just `gcNeeded()`?

GrahamCampbell commented on src/Illuminate/Session/Middleware/StartSession.php:
```diff
@@ -143,13 +144,18 @@ protected function storeCurrentUrl(Request $request, $session)
-        if ($this->configHitsLottery($config)) {
-            $session->getHandler()->gc($this->getSessionLifetimeInSeconds());
+        // Cache based sessions use their own mechanism to invalidate entries
+        // once their lifetime expires so there's no need to perform the garbage
```
could you move `garbage` onto the next line plz :P

HolgerW1 replied:
I can move the garbage wherever you want it ;-)

Commit:
Removed nesting for the session gc check
Better name for the method checking if a session handler needs garbage collection
Comment wrapping

JosephSilber commented on src/Illuminate/Session/Middleware/StartSession.php:
```diff
@@ -243,4 +252,16 @@ protected function usingCookieSessions()
+     */
+    protected function gcNeeded(SessionInterface $session)
+    {
+        return ! ($session->getHandler() instanceof CacheBasedSessionHandler);
```
We should have something more explicit than this.

GrahamCampbell replied:
don't need the bracks here either

HolgerW1 replied:
> We should have something more explicit than this.

@JosephSilber You mean sth. like a class property on the handler indicating if the handler class needs gc or not? If you mean that then yes, I would agree but I guess anything beyond the current changes would be breaking changes imo.

> don't need the bracks here either

@GrahamCampbell Imo the length of the expression followed by the `!` negation justifies the brackets because it makes the conditional statement more readable and expressive.
I understand that everyone has their own programming style and that there might be a very stringent code style guideline but it is kind of time consuming changing all those little things in PRs. Don't you have a style guide listing all those little things beyond PSR-2 ?

GrahamCampbell replied:
> Imo the length of the expression followed by the ! negation justifies the brackets because it makes the conditional statement more readable and expressive.

They are unnecessary. At the end of the day it doesn't matter though. I'll just remove them myself after merging so this matches our code style.

GrahamCampbell replied:
> Don't you have a style guide listing all those little things beyond PSR-2 ?

Yes, but we don't have fixers for this yet. StyleCI fixes 95% of our code style for us.

taylorotwell commented:
They might in the future and it doesn't hurt to call them just in case.

HolgerW1 commented:
> They might in the future and it doesn't hurt to call them just in case.

Well they **currently** don't and if they do in the future then why not introduce a method then to indicate if a session handler needs garbage collection or not? There's no need to trigger gc or play the gc lottery if the session handler doesn't support or need it.
